support/export/rmtab.c

```diff
@@ -96,3 +97,98 @@ rmtab_read(void)
 	}
 	return 0;
 }
+
+/*
+ * Check if an rmtab_path is exported given an export_path from etab file.
+ * The path is considered exported if it is an exact match or is part of
+ * a path nested under an export.
+ */
+static int
+is_exported(const char* rmtab_path, const char* export_path) {
+	/* first check that rmtab_path starts with the export_path */
+	if (strstr(rmtab_path, export_path) != rmtab_path)
+		return 0;
+
+	/* then confirm that it is an exact match or that it's nested */
+	char next = rmtab_path[strlen(export_path)];
+	return next == '\0' || next == '/';
+}
+
+/*
+ * The rmtab file can acculmulate stale entries over time. So when the list of
+ * exports changes, make a best effort attempt to insure that every entry in
+ * rmtab has a corresonding export.
+ */
+void
+rmtab_rebuild(void)
+{
+	struct rmtabent	*rep;
+	struct stat	stb;
+	FILE		*fp;
+	int		rmtab_lockid, etab_lockid;
+	int		entries_seen = 0, entries_copied = 0;
+
+	/* Hold locks for the rmtab and etab files during rebuilding */
+	if ((rmtab_lockid = xflock(_PATH_RMTABLCK, "w")) < 0)
+		return;
+
+	/* fast path for no NFSv3 mounts */
+	if (stat(_PATH_RMTAB, &stb) < 0 || stb.st_size == 0) {
+		xfunlock(rmtab_lockid);
+		return;
+	}
+
+	if ((etab_lockid = xflock(_PATH_ETABLCK, "r")) < 0) {
+		xfunlock(rmtab_lockid);
+		return;
+	}
+
+	if (!setrmtabent("r")) {
+		xfunlock(rmtab_lockid);
+		xfunlock(etab_lockid);
+		return;
+	}
+	if (!(fp = fsetrmtabent(_PATH_RMTABTMP, "w"))) {
+		endrmtabent();
+		xfunlock(rmtab_lockid);
+		xfunlock(etab_lockid);
+		return;
+	}
+
+	/* Visit all the entries in rmtab file */
+	while ((rep = getrmtabent(1, NULL)) != NULL) {
+		struct exportent *xp;
+		int saved = 0;
+
+		/* filter entries that are not exported */
+		setexportent(_PATH_ETAB, "r");
+		while ((xp = getexportent_nofollow(0, 0)) != NULL) {
+			if (is_exported(rep->r_path, xp->e_path)) {
+				fputrmtabent(fp, rep, NULL);
+				entries_copied++;
+				saved = 1;
+				break;
+			}
+		}
+		entries_seen++;
```

Comment:
Would be it be helpful ever to log the value of this and entries_copied?

Reply:
No , I don't think it is necessary.  The "purged stale rmtab entry" messages will give us a count of the entries we purged and we can infer the other counts by looking a `rmtab` file.  If we don't already collect `rmtab` in support bundle then we probably should start collecting it.